### ISSUE TITLE
AMP error propagation does not work on subclasses

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2292]
|<img alt="htj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| htj reported|
|-|-|
|Trac ID|trac#2292|
|Type|defect|
|Created|2006-12-11 15:47:03Z|

If I create one AMP command, and another to inherit from it the exception to error tag mapping is not inherited. See the attached file, and try to comment in line 54, and comment out line 53. I am not sure this is intended behavior, but it sure isn't obvious. I suspect this may be due to the meta class magic that happens in the Command class, but i am not sure.

I find the error mapping system in AMP a bit too magical (errors are wrapped quite a few places in amp.py), and there are things which cannot be done. The mapping with exceptions to tags is what the server side needs. However the client side really needs a mapping from error tags to exceptions. If I do this:
```
errors = { Exception1 : 'ERROR_TAG',
           Exception2 : 'ERROR_TAG' }
```
It is not obvious what will happen on the client side as dictionaries do not guarantee ordering. In bug [#2273](https://github.com/twisted/twisted/issues/2273) I suggested flipping them but this just flips the problem to the server side. I think it would be better if both mappings where explicitly defined, although this would create more work when creating a protocol. If this is done, it might also be possible to remove some of the metaclass stuff.

Attachments:

* [test_amp.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d0d/d0d323aaf1e89bc1b4091e6d83a993c2a06a5612/8c57458bd87c96398ab187bd5109eab104a827eb.py) (1274 bytes) - added by htj on 2006-12-11 15:48:15Z - Example showing how error mappings are not inherited in AMP commands.
* [test_amp.3.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d0d/d0d323aaf1e89bc1b4091e6d83a993c2a06a5612/0a49fbf69f4e2f7cb40ee00ac1d788cca0ce10bf.py) (1236 bytes) - added by htj on 2006-12-11 17:11:46Z - Fool-prof version of the previous file :-)
* [test_amp4.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d0d/d0d323aaf1e89bc1b4091e6d83a993c2a06a5612/66035538b4e02b3b85306ee86040794d3ba9fd5e.py) (870 bytes) - added by glyph on 2006-12-20 22:57:36Z - simplified version of the test

<details><summary>Searchable metadata</summary>

```
trac-id__2292 2292
type__defect defect
reporter__htj htj
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1165852023000000 1165852023000000
changetime__1166847289000000 1166847289000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

(I just fixed some distracting grammatical errors in the description.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set status to `closed`|
|-|-|

The test case you attached passes against current trunk...

```
glyph@alastor:~/Desktop$ trial test_amp.py
Running 1 tests.
test_amp
  AmpTestCase
    testErrorPropagation ...                                               [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 0.016s

PASSED (successes=1)
```

... so I'm closing this ticket as invalid.  Thanks for writing a unit test which made checking this very simple for me :).

Your other comments are effectively a duplicate of [#2285](https://github.com/twisted/twisted/issues/2285).  Feel free to comment there.

**Comment by twisted-trac** [comment]
|<img alt="htj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|htj set status to `reopened`|
|-|-|

Thank you for not reading the instructions on the test.

I've attached a new, which does fail on current trunk.

The comments about error propagation being a dupe of [#2285](https://github.com/twisted/twisted/issues/2285) is true. I think it was spawned due to [#2273](https://github.com/twisted/twisted/issues/2273) though. Does not matter though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

OK, I understand the error a bit better here.  The test case is still wonky, but even after correcting all of its weirdnesses it still fails.  Thanks for resubmitting.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Ready for review in ```bfnt-2292```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid set owner to @glyph|
|-|-|

 * NormalCommandProtocol, InheritedCommandProtocol, AddedCommandProtocol don't have docstrings.
 * Since the docstring at the top of twisted.protocols.amp is the entirety of AMP's documentation it might be worth mentioning:
  1. the usage of amp.Command.errors and amp.Command.fatalErrors
  1. that the expected behavior is that they be inherited (this could just as well be documented in amp.Command's docstring)
 * I like CommandInheritanceTests.errorCheck, but the extra new line after it's definition seems extraneous.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph removed owner|
|-|-|

Point-for-point:

 * I tried to add the most informative docstrings possible.  It should be good now.
 * This highlighted an additional problem which is that the core facility of AMP, i.e. ```callRemote```, was mostly just undocumented.  In fact it stopped in the middle of a sentence!
   1. ```Command.errors``` is now touched upon in the module docstring, and more thoroughly described in ```Command```'s docstring.  ```fatalErrors``` was already documented by describing how it was different than ```errors```, so I think that's adequate.
   1. The behavior of inheritance is now documented in ```Command```.
 * There are supposed to be 2 blank lines after each method.  This is from the [Divmod coding standard](http://www.divmod.org/trac/wiki/CodingStandard), which is like an even stricter interpretation of the existing Twisted standard.  I think I've adjusted the rest of the patch to conform to this now.

Ready for re-review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @glyph|
|-|-|

At the end of the first paragraph added to the amp module docstring there is "zero:" but I guess it should be "zero::" (like the following paragraphs use).

Could you name the new tests with ```test_```?

I'm not sure about the wisdom of the actual math being done in the examples, but I don't suppose it's a very big deal.

If you could replace the semi-colons with newlines in the examples, that would be nice too. :)

Everything else looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @exarkun|
|-|-|

I corrected the :: and the test names.

The examples could probably be better, but then, there could probably be a short _book_ about AMP and how to use it, and it's not even done yet... I think I'm going to leave the math.

Semi-colons were replaced before you even reassigned the review :).

Is it ready to merge?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @glyph|
|-|-|

Looks good, merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set status to `closed`|
|-|-|

(In [19011]) Fix Command inheritance in AMP, and improve documentation.

Command objects previously required all inherited commands to re-declare all
errors handled by their parent commands.  That is now taken care of
automatically by the metaclass.

This change also improves documentation of callRemote and Command, since during
the process of fixing this bug it was discovered that not only were their
inheritance semantics undocumented, the rest of their semantics were mostly
undocumented too.

Author: glyph

Reviewers: dreid, exarkun

Fixes [#2292](https://github.com/twisted/twisted/issues/2292)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
